src/data_generators/publisher_data_file.py

```diff
@@ -0,0 +1,198 @@
+        self._data = deepcopy(impression_log_data)
+        self._data.sort(key=lambda x: x[1])  # sort by spend
+        self._name = name
+        self._spend = max([spend for (_, spend) in impression_log_data])
+        self._reach = len(set([id for (id, _) in impression_log_data]))
+
+    @property
+    def impressions(self):
+        """Total number of impressions represented by this PublisherDataFile."""
+        return len(self._data)
+
+    @property
+    def spend(self):
+        """Total spend represented by this PublisherDataFile."""
+        return self._spend
+
+    @property
+    def reach(self):
+        """Total number of unique users represented by this PublisherDataFile."""
+        return self._reach
+
+    @property
+    def name(self):
+        """Returns the name associated with this PublisherDataFile."""
+        return self._name
+
+    def spend_by_impressions(self, impressions):
+        """Returns the amount spent to obtain a given number of impressions."""
+        if not impressions:
+            return 0.0
+        return self._data[impressions - 1][1]
+
+    def user_counts_by_impressions(self, impressions: int) -> Dict[int, int]:
+        """Number of times each user is reached for a given impression buy.
+
+        Args:
+          impressions:  The number of impressions that are shown (purchased).
+        Returns:
+          A dictionary D mapping user id's to frequencies.  D[u] is the number
+          of times that user u is reached for the given impression purchase.
+        """
+        return dict(Counter([id for id, _ in self._data[:impressions]]))
+
+    def user_counts_by_spend(self, spend: float) -> Dict[int, int]:
+        """Number of times each user is reached for a given spend.
+
+        Args:
+          spend:  The total spend.
+        Returns:
+          A dictionary D mapping user id's to frequencies.  D[u] is the number
+          of times that user u is reached for the given spend.
+        """
+        return dict(Counter([id for id, s in self._data if s <= spend]))
+
+    def write_publisher_data_file(self, parent_dir: str, filename: str = None) -> None:
+        """Writes this PublisherDataFile object to disk.
+
+        Args:
+          parent_dir:  The directory where the file is to be written.
+          filename:  The name of the file.  If not specified, then the name
+            given in the object constructor is used.  If no name was given
+            in the object constructor, then a random string is chosen.
+        """
+        if not filename and not self._name:
+            filename = "{:012d}".format(randint(0, 1e12))
+        elif not filename:
+            filename = self._name
+        fullpath = join(parent_dir, filename)
+        with open(fullpath, "w") as file:
+            for user_id, spend in self._data:
+                file.write("{},{}\n".format(user_id, spend))
+
+    @classmethod
+    def read_publisher_data_file(cls, filepath: str) -> "PublisherDataFile":
```

Comment:
Same here. Consider a file object instead of a path.

Reply:
Done.